File: examples/chatapp/python-be/app/gql/presence.py
from __future__ import annotations

import uuid
from collections.abc import AsyncIterator

import strawberry
from strawberry.types import Info

from .. import db
from .helpers import (
    PRESENCE_TOPIC,
    chat_topic,
    loaders,
    parse_id,
    publish_event,
    require_member,
    require_user,
    sub_events,
)
from .types import TypingEvent, User, user_from_row


@strawberry.type
class PresenceQuery:
    @strawberry.field
    async def presence(self, info: Info, user_ids: list[strawberry.ID]) -> list[User]:
        await require_user(info)
        ids = [parse_id(x) for x in user_ids]
        rows = await loaders(info)["users"].load_many(ids)
        return [user_from_row(r) for r in rows if r]


@strawberry.type
class PresenceMutation:
    @strawberry.mutation
    async def set_typing(self, info: Info, chat_id: strawberry.ID, typing: bool) -> bool:
        u = await require_user(info)
        cid = parse_id(chat_id)
        await require_member(info, cid, u["id"])
        # The indicator rides the pubsub 'typing' event; nothing reads a kv typing key.
        await publish_event(
            info,
            chat_topic(cid),
            {"type": "typing", "user_id": str(u["id"]), "typing": typing},
        )
        return True

    @strawberry.mutation
    async def heartbeat(self, info: Info) -> bool:
        u = await require_user(info)
        forge = info.context["forge"]
        await forge.kv_set(f"online:{u['id']}", "1", info.context["presence_ttl"])
        await publish_event(
            info,
            PRESENCE_TOPIC,
            {"type": "presence", "user_id": str(u["id"]), "online": True},
        )
        return True


@strawberry.type
class PresenceSubscription:
    @strawberry.subscription
    async def typing(self, info: Info, chat_id: strawberry.ID) -> AsyncIterator[TypingEvent]:
        u = await require_user(info)
        pool = info.context["pool"]
        cid = parse_id(chat_id)
        await require_member(info, cid, u["id"])
        me_id = u["id"]
        async for ev in sub_events(info, chat_topic(cid)):
            if ev.get("type") != "typing":
                continue
            uid = uuid.UUID(ev["user_id"])
            if uid == me_id:
                continue
            row = await db.users_by_ids(pool, [uid])
            if row:
                yield TypingEvent(user=user_from_row(row[0]), typing=bool(ev.get("typing")))

    @strawberry.subscription
    async def presence_changed(
        self, info: Info, user_ids: list[strawberry.ID]
    ) -> AsyncIterator[User]:
        await require_user(info)
        pool = info.context["pool"]
        wanted = {parse_id(x) for x in user_ids}
        async for ev in sub_events(info, PRESENCE_TOPIC):
            if ev.get("type") != "presence":
                continue
            uid = uuid.UUID(ev["user_id"])
            if uid not in wanted:
                continue
            row = await db.users_by_ids(pool, [uid])
            if row:
                yield user_from_row(row[0])

File: examples/chatapp/rust-be/src/gql/presence.rs
use async_graphql::{Context, ID, Object, Result, Subscription};
use futures_util::{Stream, StreamExt};

use crate::context::{Event, PRESENCE_TOPIC, chat_topic};
use crate::db;
use crate::gql::helpers::{app, guarded, map_db, me, parse_id, require_member};
use crate::gql::types::{GqlUser, TypingEvent};

#[derive(Default)]
pub struct PresenceQuery;

#[Object]
impl PresenceQuery {
    async fn presence(&self, ctx: &Context<'_>, user_ids: Vec<ID>) -> Result<Vec<GqlUser>> {
        let c = app(ctx)?;
        me(ctx)?;
        let mut ids = Vec::with_capacity(user_ids.len());
        for id in &user_ids {
            ids.push(parse_id(id)?);
        }
        Ok(db::users_by_ids(&c.pool, &ids)
            .await
            .map_err(map_db)?
            .into_iter()
            .map(GqlUser)
            .collect())
    }
}

#[derive(Default)]
pub struct PresenceMutation;

#[Object]
impl PresenceMutation {
    async fn set_typing(&self, ctx: &Context<'_>, chat_id: ID, typing: bool) -> Result<bool> {
        let c = app(ctx)?;
        let user = me(ctx)?;
        let cid = parse_id(&chat_id)?;
        require_member(c, cid, user.id).await?;
        // The indicator rides the pubsub `typing` event; nothing reads a kv key.
        c.publish(
            &chat_topic(cid),
            &Event::Typing {
                user_id: user.id,
                typing,
            },
        )
        .await;
        Ok(true)
    }

    async fn heartbeat(&self, ctx: &Context<'_>) -> Result<bool> {
        let c = app(ctx)?;
        let user = me(ctx)?;
        c.touch_presence(user.id).await.map_err(map_db)?;
        c.publish(
            PRESENCE_TOPIC,
            &Event::Presence {
                user_id: user.id,
                online: true,
            },
        )
        .await;
        Ok(true)
    }
}

#[derive(Default)]
pub struct PresenceSubscription;

#[Subscription]
impl PresenceSubscription {
    async fn typing(
        &self,
        ctx: &Context<'_>,
        chat_id: ID,
    ) -> Result<impl Stream<Item = TypingEvent> + use<>> {
        let c = app(ctx)?.clone();
        let user = me(ctx)?;
        let cid = parse_id(&chat_id)?;
        require_member(&c, cid, user.id).await?;
        let raw = c
            .forge
            .pubsub()
            .subscribe(&chat_topic(cid))
            .await
            .map_err(map_forge)?;
        let me_id = user.id;
        let raw = guarded(c.clone(), user, raw);
        Ok(raw.filter_map(move |item| {
            let c = c.clone();
            async move {
                let bytes = item.ok()?;
                match serde_json::from_slice::<Event>(&bytes).ok()? {
                    Event::Typing { user_id, typing } if user_id != me_id => {
                        let u = db::users_by_ids(&c.pool, &[user_id]).await.ok()?.pop()?;
                        Some(TypingEvent {
                            user: GqlUser(u),
                            typing,
                        })
                    }
                    _ => None,
                }
            }
        }))
    }

    async fn presence_changed(
        &self,
        ctx: &Context<'_>,
        user_ids: Vec<ID>,
    ) -> Result<impl Stream<Item = GqlUser> + use<>> {
        let c = app(ctx)?.clone();
        let user = me(ctx)?;
        let mut wanted = Vec::with_capacity(user_ids.len());
        for id in &user_ids {
            wanted.push(parse_id(id)?);
        }
        let raw = c
            .forge
            .pubsub()
            .subscribe(PRESENCE_TOPIC)
            .await
            .map_err(map_forge)?;
        let raw = guarded(c.clone(), user, raw);
        Ok(raw.filter_map(move |item| {
            let c = c.clone();
            let wanted = wanted.clone();
            async move {
                let bytes = item.ok()?;
                match serde_json::from_slice::<Event>(&bytes).ok()? {
                    Event::Presence { user_id, .. } if wanted.contains(&user_id) => {
                        db::users_by_ids(&c.pool, &[user_id])
                            .await
                            .ok()?
                            .pop()
                            .map(GqlUser)
                    }
                    _ => None,
                }
            }
        }))
    }
}

fn map_forge(e: forgelib::ForgeError) -> async_graphql::Error {
    crate::gql::helpers::map_forge(e)
}
